File: android/app/src/main/java/com/noop/data/BackupManifest.kt
package com.noop.data

import java.io.File
import java.security.MessageDigest
import org.json.JSONObject

/**
 * Shared, self-describing metadata for a `.noopbak` ZIP.
 *
 * The payload database remains engine-native (Room on Android, GRDB on Apple), but both clients use
 * these exact JSON keys. Manifest-less archives remain valid legacy backups. A current manifest lets
 * the importer reject a wrong-platform or future-schema restore before opening SQLite and verifies
 * plaintext folder snapshots with streaming SHA-256.
 */
data class BackupManifest(
    val format: String,
    val version: Int,
    val createdAtEpochMs: Long,
    val sourcePlatform: String,
    val databaseEngine: String,
    val databaseSchemaVersion: Int,
    val settingsSchemaVersion: Int?,
    val appVersion: String?,
    val payloads: Payloads,
) {
    data class Payload(val path: String, val bytes: Long, val sha256: String)
    data class Payloads(val database: Payload, val settings: Payload?)

    fun encode(): String {
        fun payloadJson(value: Payload): JSONObject = JSONObject()
            .put("path", value.path)
            .put("bytes", value.bytes)
            .put("sha256", value.sha256)

        val payloadObject = JSONObject().put("database", payloadJson(payloads.database))
        payloads.settings?.let { payloadObject.put("settings", payloadJson(it)) }
        return JSONObject()
            .put("format", format)
            .put("version", version)
            .put("createdAtEpochMs", createdAtEpochMs)
            .put("sourcePlatform", sourcePlatform)
            .put("databaseEngine", databaseEngine)
            .put("databaseSchemaVersion", databaseSchemaVersion)
            .apply {
                settingsSchemaVersion?.let { put("settingsSchemaVersion", it) }
                appVersion?.let { put("appVersion", it) }
            }
            .put("payloads", payloadObject)
            .toString()
    }

    /**
     * A user-facing incompatibility/integrity problem, or null when this manifest and the extracted
     * payloads are safe for this client. Hashes are streamed; the database is never loaded in memory.
     */
    fun validationProblem(
        databaseFile: File,
        settingsFile: File?,
        expectedDatabaseEntryName: String,
        expectedSettingsEntryName: String,
        currentPlatform: String,
        currentDatabaseEngine: String,
        currentDatabaseSchemaVersion: Int,
    ): String? {
        if (format != FORMAT_NAME) return "This file uses an unknown NOOP backup format."
        if (version != FORMAT_VERSION) {
            return "This backup uses unsupported container version $version. Update NOOP and try again."
        }
        if (createdAtEpochMs < 0L || databaseSchemaVersion <= 0) {
            return "This backup manifest contains invalid version metadata."
        }
        if (sourcePlatform !in setOf(PLATFORM_APPLE, PLATFORM_ANDROID) ||
            databaseEngine !in setOf(ENGINE_GRDB, ENGINE_ROOM)
        ) {
            return "This backup manifest contains an unknown source platform or database engine."
        }
        if (sourcePlatform != currentPlatform || databaseEngine != currentDatabaseEngine) {
            val source = if (sourcePlatform == PLATFORM_APPLE) "Apple" else "Android"
            val target = if (currentPlatform == PLATFORM_APPLE) "Apple" else "Android"
            return "This is an $source full-device backup and cannot replace the $target database. " +
                "Use NOOP's portable CSV export to move health history between platforms."
        }
        if (databaseSchemaVersion > currentDatabaseSchemaVersion) {
            return "This backup was created by a newer NOOP database schema. Update NOOP before restoring it."
        }
        if (payloads.database.path != expectedDatabaseEntryName) {
            return "This backup manifest points to an unexpected database payload."
        }
        payloadProblem(payloads.database, databaseFile)?.let {
            return "The backup database $it"
        }

        when {
            payloads.settings == null && settingsFile == null -> {
                if (settingsSchemaVersion != null) {
                    return "This backup manifest declares settings metadata without a settings payload."
                }
            }
            payloads.settings == null && settingsFile != null ->
                return "This backup contains settings that are not covered by its integrity manifest."
            payloads.settings != null && settingsFile == null ->
                return "This backup is missing the settings payload declared by its manifest."
            else -> {
                val settings = requireNotNull(payloads.settings)
                val file = requireNotNull(settingsFile)
                if (settings.path != expectedSettingsEntryName ||
                    settingsSchemaVersion == null || settingsSchemaVersion <= 0
                ) {
                    return "This backup manifest contains invalid settings metadata."
                }
                payloadProblem(settings, file)?.let {
                    return "The backup settings $it"
                }
            }
        }
        return null
    }

    companion object {
        const val ENTRY_NAME = "manifest.json"
        const val FORMAT_NAME = "noop-backup"
        const val FORMAT_VERSION = 1
        const val PLATFORM_APPLE = "apple"
        const val PLATFORM_ANDROID = "android"
        const val ENGINE_GRDB = "grdb"
        const val ENGINE_ROOM = "room"

        fun create(
            databaseFile: File,
            databaseEntryName: String,
            settingsBytes: ByteArray?,
            settingsEntryName: String,
            createdAtEpochMs: Long,
            sourcePlatform: String,
            databaseEngine: String,
            databaseSchemaVersion: Int,
            settingsSchemaVersion: Int?,
            appVersion: String?,
        ): BackupManifest {
            val database = Payload(
                path = databaseEntryName,
                bytes = databaseFile.length(),
                sha256 = sha256(databaseFile),
            )
            val settings = settingsBytes?.let {
                Payload(
                    path = settingsEntryName,
                    bytes = it.size.toLong(),
                    sha256 = sha256(it),
                )
            }
            return BackupManifest(
                format = FORMAT_NAME,
                version = FORMAT_VERSION,
                createdAtEpochMs = createdAtEpochMs,
                sourcePlatform = sourcePlatform,
                databaseEngine = databaseEngine,
                databaseSchemaVersion = databaseSchemaVersion,
                settingsSchemaVersion = if (settings == null) null else settingsSchemaVersion,
                appVersion = appVersion,
                payloads = Payloads(database, settings),
            )
        }

        fun decode(json: String): BackupManifest? = runCatching {
            fun payload(value: JSONObject): Payload = Payload(
                path = value.getString("path"),
                bytes = value.getLong("bytes"),
                sha256 = value.getString("sha256"),
            )

            val root = JSONObject(json)
            val payloads = root.getJSONObject("payloads")
            BackupManifest(
                format = root.getString("format"),
                version = root.getInt("version"),
                createdAtEpochMs = root.getLong("createdAtEpochMs"),
                sourcePlatform = root.getString("sourcePlatform"),
                databaseEngine = root.getString("databaseEngine"),
                databaseSchemaVersion = root.getInt("databaseSchemaVersion"),
                settingsSchemaVersion = if (root.has("settingsSchemaVersion")) {
                    root.getInt("settingsSchemaVersion")
                } else {
                    null
                },
                appVersion = if (root.has("appVersion")) root.getString("appVersion") else null,
                payloads = Payloads(
                    database = payload(payloads.getJSONObject("database")),
                    settings = if (payloads.has("settings") && !payloads.isNull("settings")) {
                        payload(payloads.getJSONObject("settings"))
                    } else {
                        null
                    },
                ),
            )
        }.getOrNull()

        private fun payloadProblem(payload: Payload, file: File): String? {
            if (payload.bytes < 0L || !payload.sha256.matches(Regex("[0-9a-f]{64}"))) {
                return "has invalid integrity metadata."
            }
            if (!file.isFile) return "could not be read."
            if (file.length() != payload.bytes) return "size does not match its integrity manifest."
            if (sha256(file) != payload.sha256) return "hash does not match its integrity manifest."
            return null
        }

        private fun sha256(file: File): String {
            val digest = MessageDigest.getInstance("SHA-256")
            file.inputStream().use { input ->
                val buffer = ByteArray(1_048_576)
                while (true) {
                    val count = input.read(buffer)
                    if (count < 0) break
                    digest.update(buffer, 0, count)
                }
            }
            return digest.digest().toHex()
        }

        private fun sha256(bytes: ByteArray): String =
            MessageDigest.getInstance("SHA-256").digest(bytes).toHex()

        private fun ByteArray.toHex(): String =
            joinToString(separator = "") { "%02x".format(it.toInt() and 0xff) }
    }
}

File: Packages/WhoopStore/Sources/WhoopStore/BackupManifest.swift
import CryptoKit
import Foundation

/// Self-describing metadata for a `.noopbak` container.
///
/// The SQLite payload remains native to its database engine (GRDB on Apple, Room on Android), but
/// the outer ZIP contract is shared. A manifest lets either client reject a wrong-platform or
/// future-schema restore before opening SQLite, and gives plaintext folder snapshots the same
/// payload-integrity check as encrypted manual exports. Manifest-less backups remain valid legacy
/// inputs.
public struct BackupManifest: Codable, Equatable, Sendable {
    public static let entryName = "manifest.json"
    public static let formatName = "noop-backup"
    public static let formatVersion = 1

    public enum SourcePlatform: String, Codable, Sendable {
        case apple
        case android
    }

    public enum DatabaseEngine: String, Codable, Sendable {
        case grdb
        case room
    }

    public struct Payload: Codable, Equatable, Sendable {
        public let path: String
        public let bytes: Int64
        public let sha256: String

        public init(path: String, bytes: Int64, sha256: String) {
            self.path = path
            self.bytes = bytes
            self.sha256 = sha256
        }
    }

    public struct Payloads: Codable, Equatable, Sendable {
        public let database: Payload
        public let settings: Payload?

        public init(database: Payload, settings: Payload?) {
            self.database = database
            self.settings = settings
        }
    }

    public let format: String
    public let version: Int
    public let createdAtEpochMs: Int64
    public let sourcePlatform: SourcePlatform
    public let databaseEngine: DatabaseEngine
    public let databaseSchemaVersion: Int
    public let settingsSchemaVersion: Int?
    public let appVersion: String?
    public let payloads: Payloads

    public init(
        format: String = BackupManifest.formatName,
        version: Int = BackupManifest.formatVersion,
        createdAtEpochMs: Int64,
        sourcePlatform: SourcePlatform,
        databaseEngine: DatabaseEngine,
        databaseSchemaVersion: Int,
        settingsSchemaVersion: Int?,
        appVersion: String?,
        payloads: Payloads
    ) {
        self.format = format
        self.version = version
        self.createdAtEpochMs = createdAtEpochMs
        self.sourcePlatform = sourcePlatform
        self.databaseEngine = databaseEngine
        self.databaseSchemaVersion = databaseSchemaVersion
        self.settingsSchemaVersion = settingsSchemaVersion
        self.appVersion = appVersion
        self.payloads = payloads
    }

    public static func make(
        databaseAt databaseURL: URL,
        databaseEntryName: String,
        settingsData: Data?,
        settingsEntryName: String,
        createdAtEpochMs: Int64,
        sourcePlatform: SourcePlatform,
        databaseEngine: DatabaseEngine,
        databaseSchemaVersion: Int,
        settingsSchemaVersion: Int?,
        appVersion: String?
    ) throws -> BackupManifest {
        let database = try payload(fileAt: databaseURL, path: databaseEntryName)
        let settings = settingsData.map {
            Payload(path: settingsEntryName, bytes: Int64($0.count), sha256: sha256($0))
        }
        return BackupManifest(
            createdAtEpochMs: createdAtEpochMs,
            sourcePlatform: sourcePlatform,
            databaseEngine: databaseEngine,
            databaseSchemaVersion: databaseSchemaVersion,
            settingsSchemaVersion: settings == nil ? nil : settingsSchemaVersion,
            appVersion: appVersion,
            payloads: Payloads(database: database, settings: settings)
        )
    }

    public func encoded() throws -> Data {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys, .withoutEscapingSlashes]
        return try encoder.encode(self)
    }

    public static func decoded(from data: Data) throws -> BackupManifest {
        try JSONDecoder().decode(BackupManifest.self, from: data)
    }

    /// Returns a user-facing incompatibility/integrity problem, or `nil` when the manifest and
    /// extracted payloads are valid for this client. This performs bounded streaming SHA-256 reads;
    /// the database is never loaded into memory.
    public func validationProblem(
        databaseAt databaseURL: URL,
        settingsAt settingsURL: URL?,
        expectedDatabaseEntryName: String,
        expectedSettingsEntryName: String,
        currentPlatform: SourcePlatform,
        currentDatabaseEngine: DatabaseEngine,
        currentDatabaseSchemaVersion: Int
    ) -> String? {
        guard format == Self.formatName else {
            return "This file uses an unknown NOOP backup format."
        }
        guard version == Self.formatVersion else {
            return "This backup uses unsupported container version \(version). Update NOOP and try again."
        }
        guard createdAtEpochMs >= 0, databaseSchemaVersion > 0 else {
            return "This backup manifest contains invalid version metadata."
        }
        guard sourcePlatform == currentPlatform, databaseEngine == currentDatabaseEngine else {
            let source = sourcePlatform == .apple ? "Apple" : "Android"
            let target = currentPlatform == .apple ? "Apple" : "Android"
            return "This is a \(source) full-device backup and cannot replace the \(target) database. Use NOOP's portable CSV export to move health history between platforms."
        }
        guard databaseSchemaVersion <= currentDatabaseSchemaVersion else {
            return "This backup was created by a newer NOOP database schema. Update NOOP before restoring it."
        }
        guard payloads.database.path == expectedDatabaseEntryName else {
            return "This backup manifest points to an unexpected database payload."
        }
        if let problem = Self.payloadProblem(payloads.database, fileAt: databaseURL) {
            return "The backup database \(problem)"
        }

        switch (payloads.settings, settingsURL) {
        case (nil, nil):
            guard settingsSchemaVersion == nil else {
                return "This backup manifest declares settings metadata without a settings payload."
            }
        case (nil, .some):
            return "This backup contains settings that are not covered by its integrity manifest."
        case (.some, nil):
            return "This backup is missing the settings payload declared by its manifest."
        case let (.some(settings), .some(url)):
            guard settings.path == expectedSettingsEntryName,
                  let settingsSchemaVersion, settingsSchemaVersion > 0 else {
                return "This backup manifest contains invalid settings metadata."
            }
            if let problem = Self.payloadProblem(settings, fileAt: url) {
                return "The backup settings \(problem)"
            }
        }
        return nil
    }

    private static func payload(fileAt url: URL, path: String) throws -> Payload {
        let attributes = try FileManager.default.attributesOfItem(atPath: url.path)
        guard let bytes = (attributes[.size] as? NSNumber)?.int64Value, bytes >= 0 else {
            throw CocoaError(.fileReadUnknown)
        }
        return Payload(path: path, bytes: bytes, sha256: try sha256(fileAt: url))
    }

    private static func payloadProblem(_ payload: Payload, fileAt url: URL) -> String? {
        guard payload.bytes >= 0, isSHA256(payload.sha256) else {
            return "has invalid integrity metadata."
        }
        guard let attributes = try? FileManager.default.attributesOfItem(atPath: url.path),
              let bytes = (attributes[.size] as? NSNumber)?.int64Value else {
            return "could not be read."
        }
        guard bytes == payload.bytes else {
            return "size does not match its integrity manifest."
        }
        guard let digest = try? sha256(fileAt: url), digest == payload.sha256 else {
            return "hash does not match its integrity manifest."
        }
        return nil
    }

    private static func sha256(fileAt url: URL) throws -> String {
        let handle = try FileHandle(forReadingFrom: url)
        defer { try? handle.close() }
        var hash = SHA256()
        while let chunk = try handle.read(upToCount: 1_048_576), !chunk.isEmpty {
            hash.update(data: chunk)
        }
        return hash.finalize().map { String(format: "%02x", $0) }.joined()
    }

    private static func sha256(_ data: Data) -> String {
        SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
    }

    private static func isSHA256(_ value: String) -> Bool {
        value.count == 64 && value.allSatisfy { $0.isHexDigit && !$0.isUppercase }
    }
}
